src/main/java/org/opensearch/ad/transport/handler/AnomalyIndexHandler.java

```diff
@@ -95,45 +93,50 @@ public AnomalyIndexHandler(
                 AnomalyDetectorSettings.MAX_RETRY_FOR_BACKOFF.get(settings)
             );
         this.indexName = indexName;
-        this.createIndex = createIndex;
-        this.indexExists = indexExists;
+        this.anomalyDetectionIndices = anomalyDetectionIndices;
         this.fixedDoc = false;
         this.clientUtil = clientUtil;
         this.indexUtils = indexUtils;
         this.clusterService = clusterService;
     }
 
-    /**
-     * Since the constructor needs to provide injected value and Guice does not allow Boolean to be there
-     * (claiming it does not know how to instantiate it), caller needs to manually set it to true if
-     * it want to save to a specific doc.
-     * @param fixedDoc whether to save to a specific doc Id
-     */
-    public void setFixedDoc(boolean fixedDoc) {
-        this.fixedDoc = fixedDoc;
-    }
-
-    public void index(T toSave, String detectorId) {
+    // TODO: check if user has permission to index.
+    public void index(T toSave, String detectorId, String customIndexName) {
         if (indexUtils.checkIndicesBlocked(clusterService.state(), ClusterBlockLevel.WRITE, this.indexName)) {
             LOG.warn(String.format(Locale.ROOT, CANNOT_SAVE_ERR_MSG, detectorId));
             return;
         }
 
         try {
-            if (!indexExists.getAsBoolean()) {
-                createIndex
-                    .accept(ActionListener.wrap(initResponse -> onCreateIndexResponse(initResponse, toSave, detectorId), exception -> {
-                        if (ExceptionsHelper.unwrapCause(exception) instanceof ResourceAlreadyExistsException) {
-                            // It is possible the index has been created while we sending the create request
-                            save(toSave, detectorId);
-                        } else {
-                            throw new AnomalyDetectionException(
-                                detectorId,
-                                String.format(Locale.ROOT, "Unexpected error creating index %s", indexName),
-                                exception
-                            );
-                        }
-                    }));
+            if (customIndexName != null) {
+                // Only create custom AD result index when create detector, won’t recreate custom AD result index in realtime
+                // job and historical analysis later if it’s deleted. If user delete the custom AD result index, and AD plugin
+                // recreate it, that may bring confusion.
+                if (!anomalyDetectionIndices.doesIndexExist(customIndexName)) {
+                    throw new EndRunException(detectorId, CAN_NOT_FIND_RESULT_INDEX + customIndexName, true);
+                }
+                if (!anomalyDetectionIndices.isValidResultIndex(customIndexName)) {
```

Comment:
What happens if an index pattern tries to be set here / some name with a wildcard? Tried to test on your test domain but seems to be broken.

Reply:
Index name should not contain "*", OpenSearch will fail to create index with name including "*"